xarray/core/dataset.py

```diff
@@ -9293,6 +9294,12 @@ def pad(
             if not pad_dims.intersection(xindexes.get_all_dims(k)):
                 indexes[k] = idx
 
+        per_data_var_constant_values = {}
+        if isinstance(constant_values, dict):
```

Comment:
The typing claims that any mapping works, but here you are only checking dicts.
I would propose to change to `utils.is_dict_like` and find an alternative to pop (Mapping does not define pop because it is read-only). The simplest way would be to transform it into a dict first.

Reply:
That's a fair point, to be honest I just followed the way it is done in the implementation of `pad`  for the variables in `variable.py`. Maybe I should just change all these type hints to `dict` rather than `Mapping`. What do you think?

Reply:
From a typing perspective dict is always problematic because it is invariant.
I would try to stick with Mapping instead.

Reply:
Ah okay, so just change this one `isinstance(constant_values, dict)` to `utils.is_dict_like(constant_values)` then?

Reply:
Yes and no.
For the check it should suffice, but the general Mapping class has no `pop` defined because it is read-only (non-mutable).
Anyway, I guess the user would find it weird that this input is changed in-place, so remove the `pop` method and replace it with a get item call (not sure if the Mapping ABC defines a `get` method).

Reply:
I wonder if the check needs to be more thorough? It is totally possible to create `Dataset` objects with a data variable that has the same name as a dimension:
```python
xr.Dataset({"x": ("x", [1, 2])}).drop_indexes("x").reset_coords("x")
```
so we could potentially break the original use case if we remove all data variables from the dictionary, because we could also remove a dimension name.
---
In general, I believe we need to be thorough in defining the semantics of the new feature:
- should we also allow padding coordinate variables?
- if we use it, should it still be possible to specify blanket dimension padding? I.e. should it be possible to mix the two? If so, how do we figure out which is which? By type?
- what should happen if we specify padding only for a subset of the variables? Dimensions need to have the same size for all variables in a `Dataset`, so either we need to specify a fallback (dimension names), or we need to require specifying the padding for all affected variables (and raise otherwise).

Reply:
> Yes and no. For the check it should suffice, but the general Mapping class has no `pop` defined because it is read-only (non-mutable). Anyway, I guess the user would find it weird that this input is changed in-place, so remove the `pop` method and replace it with a get item call (not sure if the Mapping ABC defines a `get` method).

That's a very fair point, always forget that about dicts :upside_down_face: , I'll figure out a different way to do it.

Reply:
Alrighty, changed, now the input `constant_values` is not mutated in place.

Reply:
> I wonder if the check needs to be more thorough? It is totally possible to create `Dataset` objects with a data variable that has the same name as a dimension:
> 
> ```python
> xr.Dataset({"x": ("x", [1, 2])}).drop_indexes("x").reset_coords("x")
> ```
> 
> so we could potentially break the original use case if we remove all data variables from the dictionary, because we could also remove a dimension name.
> 
> In general, I believe we need to be thorough in defining the new feature:
> 
> * should we also allow padding coordinate variables?
> * if we use it, should it still be possible to specify blanket dimension padding? I.e. should it be possible to mix the two? If so, how do we figure out which is which? Check if the value is a `dict` and decide it's the padding for a specific variable?
> * what should happen if we specify padding values only for a subset of the variables? Dimensions need to have the same size for all variables in a `Dataset`, so either we need to specify a fallback (by dimension names), or we need to require specifying the padding for all affected variables (and raise otherwise).

It is a fair point, I have rewritten this section so the original `constant_variables` mapping is not mutated.
Regarding the bullet points:
* As I understand it was already possible to pad coordinate variables.
* As it is codded now if a `data var` has a `dim` with the same name and the user sets a value with this name in the `constant_values` then the `data var` will be padded with those values (regardless of the dim being padded) and all other `data vars` that also contain that `dim` will be padded with that value, provided that it is the `dim` being padded (a.k.a. the dim provided in `pad_width`). In the tests I also show that one can mix `data var` and `dim` in the same dict. In that case the value for the `data var` has priority over the value of the `dim`.
* I realized that it was already implemented that if one didn't set values for all `dims` that are being padded then they are be padded with 0, so I kept that. For example, if one sets a `constant_value` for `var1` and is padding a `dim` that is contained in in other `vars`, `var1` will be padded with that value along the `dim` while all other `vars` will be padded with 0 along the same `dim`.

I hope I didn't make it too confusing. Let me know if it answers your questions and if the logic makes sense :D